core/src/main/scala/scalafix/rewrite/Xor2Either.scala

```diff
@@ -1,28 +1,55 @@
 package scalafix.rewrite
 
+import scala.collection.immutable.{ListMap, Seq}
 import scala.meta.parsers.Parse
 import scala.{meta => m}
-import scalafix.util.Patch
+import scalafix.util._
+import PartialFunction._
+import scala.meta.Type.Select
+import scala.meta._
+
+class Desugared[T <: Tree: Parse](implicit semantic: SemanticApi) {
+  def unapply(original: T): Option[T] = semantic.desugared(original)
+}
 
 case object Xor2Either extends Rewrite {
   override def rewrite(ast: m.Tree, ctx: RewriteCtx): Seq[Patch] = {
-    import scala.meta._
-    val semantic = getSemanticApi(ctx)
-    class Desugared[T <: Tree: Parse] {
-      def unapply(original: T): Option[T] = semantic.desugared(original)
-    }
+    implicit val semanticApi: SemanticApi = getSemanticApi(ctx)
     object DType extends Desugared[Type]
     object DTerm extends Desugared[Term]
-    // NOTE. This approach is super inefficient, since we run semantic.desugar on
-    // every case for every node in the tree. Ideally, we first match on the
-    // syntax structure we want and then run semantic.desugar.
-    ast.collect {
-      case t @ DType(t"cats.data.Xor") =>
-        Patch(t.tokens.head, t.tokens.last, s"Either")
-      case t @ DTerm(q"cats.data.Xor.Right.apply[..$_]") =>
-        Patch(t.tokens.head, t.tokens.last, s"Right")
-      case t @ DTerm(q"cats.data.Xor.Left.apply[..$_]") =>
-        Patch(t.tokens.head, t.tokens.last, s"Left")
-    }
+
+    val typeChanger = new ChangeType(ast, ctx)
+    val methodChanger = new ChangeMethod(ast, ctx)
+    val importAdder = new AddImport(ast, ctx)
+
+    //Create a sequence of type replacements
+    val replacementTypes = List(
+      ReplaceType(t"cats.data.XorT", t"cats.data.EitherT", "EitherT"),
+      ReplaceType(t"cats.data.Xor", t"scala.util.Either", "Either"),
+      ReplaceType(t"cats.data.Xor.Left", t"scala.util.Left", "Left"),
+      ReplaceType(t"cats.data.Xor.Right", t"scala.util.Either.Right", "Right")
+    )
```

Comment:
I'm sad to hear this is a productivity killer.

You don't have to use quasiquotes. I'm guilty of using strings to strip of a prefix package name just because I just wanted to see something work, see [here](https://github.com/scalacenter/scalafix/blob/e2f97d651cbba38a287d86543fd36f9dbddbadda/scalafix-nsc/src/main/scala/scalafix/nsc/NscSemanticApi.scala#L46). However, string manipulation is almost like programming with void pointers in C, try to avoid it when possible.

An alternative to quasiquotes that's not string manipulation is using the regular Tree ADT. I do this myself all the time ([example](https://github.com/scalacenter/scalafix/blob/e2f97d651cbba38a287d86543fd36f9dbddbadda/scalafix-nsc/src/main/scala/scalafix/nsc/NscSemanticApi.scala#L31)). Quasiquotes tend to hide the fact that you're working with tree structures. With scalafmt, I found the relationship counter-intuitive between the code on the surface and the underlying parsed AST, for example

```scala
@ q"a.b.c.d".structure
res20: String = """
Term.Select(Term.Select(Term.Select(Term.Name("a"), Term.Name("b")), Term.Name("c")), Term.Name("d"))
"""
// note `d` is at the top and `a` is at the bottom
```

If you want to use quasiquotes, you might be looking for `transform` in combination with `StructurallyEqual` or `.structure`
```scala
@ val prefix = q"a.b.c"
prefix: Term.Select = a.b.c
@ val fqn = q"$prefix.d.e"
fqn: Term.Select = a.b.c.d.e
@ fqn.transform {
  case q"$somePrefix.$name" if somePrefix.structure == prefix.structure  => name
}
res15: Tree = d.e
```